imcache.go

```diff
@@ -259,42 +259,46 @@ func (c *Cache[K, V]) GetOrSet(key K, val V, exp Expiration) (value V, present b
 	}
 	// Make sure that the shard is initialized.
 	c.init()
-	currentEntry, ok := c.m[key]
+	currentNode, ok := c.m[key]
 	if !ok {
-		newEntry.node = c.queue.AddNew(key)
-		c.m[key] = newEntry
+		newEntry := entry[K, V]{key: key, val: val}
+		exp.apply(&newEntry.exp)
+		newEntry.SetDefaultOrNothing(now, c.defaultExp, c.sliding)
+		c.m[key] = c.queue.add(newEntry)
 		if c.maxEntriesLimit <= 0 || c.len() <= c.maxEntriesLimit {
 			c.mu.Unlock()
 			return val, false
 		}
-		lruNode := c.queue.Pop()
+		evictedNode := c.queue.pop()
+		evictedEntry := evictedNode.entry()
 		if c.onEviction == nil {
-			delete(c.m, lruNode.key)
+			delete(c.m, evictedEntry.key)
 			c.mu.Unlock()
 			return val, false
 		}
-		lruEntry := c.m[lruNode.key]
-		delete(c.m, lruNode.key)
+		delete(c.m, evictedEntry.key)
 		c.mu.Unlock()
-		if lruEntry.HasExpired(now) {
-			go c.onEviction(lruNode.key, lruEntry.val, EvictionReasonExpired)
+		if evictedEntry.HasExpired(now) {
+			go c.onEviction(evictedEntry.key, evictedEntry.val, EvictionReasonExpired)
 		} else {
-			go c.onEviction(lruNode.key, lruEntry.val, EvictionReasonMaxEntriesExceeded)
+			go c.onEviction(evictedEntry.key, evictedEntry.val, EvictionReasonMaxEntriesExceeded)
 		}
 		return val, false
 	}
-	c.queue.Remove(currentEntry.node)
+	currentEntry := currentNode.entry()
 	if !currentEntry.HasExpired(now) {
+		c.queue.touch(currentNode)
 		if currentEntry.HasSlidingExpiration() {
 			currentEntry.SlideExpiration(now)
-			c.m[key] = currentEntry
+			currentNode.setEntry(currentEntry)
 		}
-		c.queue.Add(currentEntry.node)
 		c.mu.Unlock()
 		return currentEntry.val, true
 	}
-	newEntry.node = c.queue.AddNew(key)
-	c.m[key] = newEntry
+	newEntry := entry[K, V]{key: key, val: val}
```

Comment:
The entry should be added if there is no entry or the existing entry is expired. It shouldn't be handled separately.